### ISSUE TITLE
ArgumentError: invalid base64 when update Rails v5.2.5

### ISSUE THREAD

**Issue by taketo1113** [rails/rails#41783]
### Steps to reproduce

When update rails v5.2.4.5 -> v5.2.5 in production env, Error occurs due to change of CSRF Token format from base64_strict to base64_urlsafe.

Error messages:
```
ArgumentError: invalid base64
File /usr/local/lib/ruby/2.6.0/base64.rb line 74 in unpack1
File /usr/local/lib/ruby/2.6.0/base64.rb line 74 in strict_decode64
File /var/www/app/vendor/bundle/ruby/2.6.0/gems/actionpack-5.2.4.5/lib/action_controller/metal/request_forgery_protection.rb line 406 in real_csrf_token
File /var/www/app/vendor/bundle/ruby/2.6.0/gems/actionpack-5.2.4.5/lib/action_controller/metal/request_forgery_protection.rb line 423 in csrf_token_hmac
File /var/www/app/vendor/bundle/ruby/2.6.0/gems/actionpack-5.2.4.5/lib/action_controller/metal/request_forgery_protection.rb line 417 in global_csrf_token
File /var/www/app/vendor/bundle/ruby/2.6.0/gems/actionpack-5.2.4.5/lib/action_controller/metal/request_forgery_protection.rb line 321 in masked_authenticity_token
File /var/www/app/vendor/bundle/ruby/2.6.0/gems/actionpack-5.2.4.5/lib/action_controller/metal/request_forgery_protection.rb line 308 in form_authenticity_token
File /var/www/app/vendor/bundle/ruby/2.6.0/gems/actionpack-5.2.4.5/lib/abstract_controller/helpers.rb line 67 in form_authenticity_token
File /var/www/app/vendor/bundle/ruby/2.6.0/gems/actionview-5.2.4.5/lib/action_view/helpers/csrf_helper.rb line 26 in csrf_meta_tags
```

### Expected behavior

In rails v5.2.x or v6.0.x, CSRF Token format should use base64_strict.
or as well as rails v6.1.x, it should switch to use `base64_strict` with option of `action_controller.urlsafe_csrf_tokens`  in rails v5.2.x or v6.0.x

Because it couldn't switch CSRF Token format, error occurs when update rails v5.2.4.5 -> v5.2.5
And also, when update rails 5.2.5 -> 6.0.x, will be raise error again due to change of CSRF Token format from base64_urlsafe (rails v5.2.5) to  base64_strict (rails v6.0.x).

### Actual behavior

 
In rails v5.2.4.5 or previous versions: CSRF Token format use `base64_strict`.
In rails v5.2.5: CSRF Token format use `base64_urlsafe`.
In rails v6.0.x, CSRF Token format use `base64_strict`.
In rails v6.1.x, CSRF Token format use `base64_urlsafe` by defalut, and it can switch to use `base64_strict` with option of `action_controller.urlsafe_csrf_tokens`

### System configuration
**Rails version**: 5.2.5
**Ruby version**: 2.7.2

**Comment by springerigor** [comment]
It's a pity, that 5.2.5 release note does not mention anything about this breaking change 😿  (https://github.com/rails/rails/compare/v5.2.4.5...v5.2.5#diff-5f81b06a0e3051b576daee16c960b21e715a6cc26d97d020c546d2fa697c9bc6)

**Comment by mohsen3** [comment]
CSRF token are refreshed with every page load, right? So, this happens if a form is loaded from a v5.2.4.5 server but submitted to a v5.2.5? I think the change is also not backwards compatible, so reverting the change also causes the same problem since some of the CSRF tokens are generated with the new format and cannot be decoded with the old one. I think the error goes away with a page refresh.

**Comment by mohsen3** [comment]
I did some experiment and it seems that the old tokens (encoded using `Base64.strict_encode64`) can always be decoded using the new method (`Base64.urlsafe_decode64`), but it's not the case for the other way around (i.e., the ones encoded with `urlsafe_encode64` cannot be decoded using `strict_decode64`).

This is a bit random and depends on the encoded string having any `/` in it, but it's true for most of the cases.

Therefore, my guess is that whoever coded that, assumed that this code is backwards compatible and there would be no issues.
BUT ... the code is not forward compatible as I explained above.

In our scenario, we have many pods running our rails app and a rolling upgrade happens at the time of deployment. So, we can get a mixture of old and new versions. That mixture resulted in a few errors but since we were worried about those errors, we rolled back the deploy that resulted in even more errors since now most of the tokens are the new version and all of our pods are running the old version. I just guess, if we let the deploy be completed, errors would have gone away way faster!

This is just a guess based on a quick investigation I did and I could be totally wrong.

**Comment by taketo1113** [comment]
@mohsen3  Thank you for your experiment.

I happened a few hundreds of errors on my app, when update rails v5.2.4.5 -> v5.2.5 in production env,
The size of the app is about 30,000 PV/hour in page views.

> so reverting the change also causes the same problem since some of the CSRF tokens are generated with the new format and cannot be decoded with the old one.

exactly right.

Then I think it's better to use the option to switch to base64_strict or base64_urlsafe in rails v5.2.x and v6.0.x as well as rails v6.1.x.

**Comment by rafaelfranca** [comment]
Just the option will not help since as soon you change its value it will raise an exception. I think we need to change 5.2 and 6.0 to be forward compatible with `base64_urlsafe` and `base64_strict`.

**Comment by rafaelfranca** [comment]
@mohsen3 @taketo1113 are any of you interested in taking a stab on making the code forward and backwards compatible? and even introduce the option?

**Comment by kamipo** [comment]
Sorry for the bothering you guys.

This compatibility issue is unintentional, so I've made a PR #41797 as a bug fix.

@rafaelfranca Is #41797 with a CHANGELOG entry acceptable as a bug fix?

**Comment by taketo1113** [comment]
@kamipo Thank you for making a PR.

Although, I think need the option to switch to `base64_strict` or `base64_urlsafe` in rails v5.2.x and v6.0.x as well as rails v6.1.x.
it doesn't have backward compatibility for v5.2.5 which only fixes to use `base64_urlsafe` to `base64_strict`.

Because the problem is that has been released to change csrf token format to use `base64_urlsafe` in rails 5.2.5.
I seem that it will occur the same error again when update next rails 5.2.x include the PR #41797, by reason of changing `base64_urlsafe` in rails v5.2.5, to `base64_strict` in PR #41797.

@rafaelfranca @kamipo What do you think about the option switching to `base64_strict` or `base64_urlsafe` to be implemented like bellow code, as well as rails v6.1.x?
https://github.com/rails/rails/blob/v6.1.3.1/railties/lib/rails/application/configuration.rb#L181

I think the approche using an option which has forward and backwards compatible.

**Comment by rafaelfranca** [comment]
I don't think the option alone would make it forward /backward compatbile. In our application we got this same error even using the configuration. We also need to change the code to be able to read both formats and introduce the config.

**Comment by unkn0wn-developer** [comment]
Updated to resolve mimemagic crisis. Now, everything is broken again. Good god.

**Comment by unkn0wn-developer** [comment]
Hi, any updates on this? We would like to fix our deployments by upgrading to 5.2.5 and remove mimemagic but this is holding us back since I got the same problem on dev and test environments. And in my case, page refresh didn't resolve this. 

Only way I could access to the website was using an incognito tab which we cannot expect our users to do that.

**Comment by unkn0wn-developer** [comment]
We tried adding the following and it worked.

-----

**_app/helpers/application_helper.rb_**
```
  def rescued_csrf_meta_tags
    csrf_meta_tags
  rescue ArgumentError
    request.reset_session
    csrf_meta_tags
  end
```
-----
**_html head tag_**
```
  = rescued_csrf_meta_tags
```
instead of "csrf_meta_tags"

-----
Not a good solution because we will eventually have to remove this as it's super unnecessary after the problem is fixed. However, it is another way to handle it.

**And I have to emphasize one more time that simply "refreshing the Page" doesn't fix Rails 5.2.5 problem for "csrf_meta_tags". And all the pages become inaccessible with Error 500. **

"refreshing the page" might work for some forms though. 

The user keeps generating errors in the background and can NEVER access the website if the user doesn't clear "cookies, site data, and cache". The user can also use incognito browser window but none of these are normal use cases.

**Comment by kamipo** [comment]
Addressed by #41797.
